### PR TITLE
Assembly changes

### PR DESCRIPTION
**Pull request by der-fruhling**
A bunch of fun asm changes:
* Add nasm files for mac and linux (in my opinion, easier to understand)
* Mark the ones that will only compile on mac as so

### CODE REVIEW

File: hello.bat [hunk 1]
```diff
@@ -0,0 +1,11 @@
+@echo off
+
+rem Windows batch script
+
+rem This is Hello World
+rem written in batch
```
Comment (jeremycole) [inline hello.bat]:
```suggestion
rem written in batch script.
```

File: hello.bat [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+for /L %%I in (1, 1, 10) do (
+    echo Hello World
+    rem Greetings
+)
```
Comment (jeremycole) [inline hello.bat]:
```suggestion
)

```

(Just end the file with a newline.)

File: hello.linux.nasm [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        mov edx, dword [helloworld_size] ; The string's size
+        int 80h ; interrupt!
+
+        dec byte [i] ; Make i go down 1
+        jmp .loop ; back to _start.loop
+.exit:
+        mov eax, 1 ; system call (int80h) for exit
+        mov ebx, 0 ; exit with success
+        int 80h ; interrupt!
+        ; the program is no longer running now
```
Comment (jeremycole) [inline hello.linux.nasm]:
```suggestion
        ; the program is no longer running now
        
```

File: hello.bat [hunk 4]
```diff
@@ -0,0 +1,11 @@
+@echo off
+
+rem Windows batch script
+
+rem This is Hello World
+rem written in batch
+
+for /L %%I in (1, 1, 10) do (
+    echo Hello World
```
Comment (jeremycole) [inline hello.bat]: Can you make this message match the one used everywhere else `Hello, World!`? It looks like [batch script has some weird quoting rules](https://www.robvanderwoude.com/escapechars.php) but it should be possible to get it to work, maaaaybe this (but you'll need to try it and see if it outputs what you want):

```suggestion
    echo Hello^, World^^!
```

File: hello.bat [hunk 5]
```diff
@@ -0,0 +1,11 @@
+@echo off
```
Comment (jeremycole) [inline hello.bat]: I think you should move batch script to a separate pull request from assembly!

File: hello.s [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Assembly (x86-64, gas)
+# Assembly (macho64, gas), only compiles on MacOS
```
Comment (jeremycole) [inline hello.s]: I don't think this description is really right – Mach-O 64 is an executable file format, not an architecture. I should think that this code is x86-64 specific and gas-specific, and probably relies on stdlib for the `_write` syscall wrapper. I am not sure that it's MacOS specific – could you clarify what's specific to Mac OS here?

I think we should make a directory for `assembly` and put all these in there with nice names, like `assembly/hello_x86-64_macos.gas` or so...

I think there are a bunch more we could end up adding. 😄

File: hello.linux.nasm [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+;  written in (nasm) Assembly.
+
+        section .bss
+
+        i: resb 1
+
+        section .rodata
+
+        helloworld_msg: db `Hello, World!\n`
+        helloworld_size: dd 14 ; Hard coded, may be improved
```
Comment (jeremycole) [inline hello.linux.nasm]: This is good – for consistency sake perhaps we should do the same thing you're doing here in the `gas` version as well?

File: hello.linux.nasm [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+        section .text
+        global _start
+
+_start: ; Entrypoint
+        mov byte [i], 10 ; set i = 10
+.loop: ; Loop until i == 0
+        cmp byte [i], 0 ; is i == 0 yet?
+        je .exit ; if so jump to _start.exit
```
Comment (jeremycole) [inline hello.linux.nasm]: This isn't really `_start.exit` is it? I think it's just a label named `.exit`? I think the names are still global, but you tell me, I am definitely ready to be wrong. ;)

File: hello.linux.nasm [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        mov eax, 4 ; system call (int80h) for write
+        mov ebx, 1 ; stdout
+        mov ecx, helloworld_msg ; The string to print
+        mov edx, dword [helloworld_size] ; The string's size
+        int 80h ; interrupt!
+
+        dec byte [i] ; Make i go down 1
```
Comment (jeremycole) [inline hello.linux.nasm]:
```suggestion
        dec byte [i] ; Subtract 1 from i
```

**Review by jeremycole — CHANGES_REQUESTED** [review]
Yay! I like new things and it looks like you're doing good work here!!! I still need to read some of the assembly but I went ahead and left a few comments. The highest level ones are probably to move the batch version to another separate PR.

File: hello.linux.nasm (hunk 7 above)
Comment (der-fruhling) [reply hello.linux.nasm]: porbably

File: hello.linux.nasm (hunk 8 above)
Comment (der-fruhling) [reply hello.linux.nasm]:
```assembly
cmp byte [i], 0
je hello!_start.exit
; ...
jmp hello!_start.loop
```

says the great ```edb``` command

File: hello.bat (hunk 4 above)
Comment (der-fruhling) [reply hello.bat]: ```echo Hello^, World^! ``` outputs what you're looking for.
I also noticed that older versions of Windows may not support the LF ending. Maybe CRLF is better.

File: hello.linux.nasm (hunk 8 above)
Comment (der-fruhling) [reply hello.linux.nasm]: Fixed for understandability.

**Comment by der-fruhling** [comment]
i think this pr has been forgotten about :cry:

**Review by jeremycole — APPROVED** [review]
🚀

**Pull request merged**